backend/yubikey.go

```diff
@@ -0,0 +1,206 @@
+package backend
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"github.com/foxboron/sbctl/config"
+	"github.com/foxboron/sbctl/logging"
+	"math/big"
+	"strings"
+	"time"
+
+	"github.com/tomis007/piv-go/v2/piv"
+)
+
+type Yubikey struct {
+	keytype BackendType
+	cert    *x509.Certificate
+	privkey *piv.KeyRSA
+}
+
+func NewYubikeyKey(conf *config.YubiConfig, desc string) (*Yubikey, error) {
+	var pub crypto.PublicKey
+	var priv crypto.PrivateKey
+	if conf.Priv == nil && conf.Pub == nil {
+		// List all smartcards connected to the system.
+		cards, err := piv.Cards()
+		if err != nil {
+			return nil, err
+		}
+
+		// Find a YubiKey and open the reader.
+		var yk *piv.YubiKey
+		for _, card := range cards {
+			if strings.Contains(strings.ToLower(card), "yubikey") {
+				if yk, err = piv.Open(card); err != nil {
+					return nil, err
+				}
+			}
+		}
+		if yk == nil {
+			return nil, fmt.Errorf("yubikey key not found")
+		}
+		conf.YK = yk
+
+		keyInfo, err := yk.KeyInfo(piv.SlotSignature)
+		if err != nil {
+			return nil, err
+		}
+		if keyInfo.PublicKey != nil {
+			if keyInfo.Algorithm == piv.AlgorithmRSA2048 {
+				logging.Println("RSA 2048 Key exists in Yubikey PIV Signature Slot, using the existing key")
+				pub = keyInfo.PublicKey
+			} else {
+				return nil, fmt.Errorf("non RSA2048 key already in the slot")
+			}
+		} else {
+			// Generate a private key on the YubiKey.
+			key := piv.Key{
+				Algorithm:   piv.AlgorithmRSA2048,
+				PINPolicy:   piv.PINPolicyAlways,
+				TouchPolicy: piv.TouchPolicyAlways,
+			}
+			logging.Println("Creating key... please press Yubikey")
+			pub, err = yk.GenerateKey(piv.DefaultManagementKey, piv.SlotSignature, key)
+			if err != nil {
+				return nil, err
+			}
+
+		}
+
+		// TODO prompt for PIN when it's not default
```

Comment:
This can probably be left as a TODO until I figure out the PIN/password support.

Reply:
I think here I'll want it to prompt for the Yubikey PIV pin when it's not `123456`. Were you talking about a different PIN?

Reply:
I'm thinking generally how to deal with password prompts in sbctl. It's not a thing yet and that needs to be planned a bit.

Reply:
I ended up using `https://github.com/manifoldco/promptui` in my yubikey backend file, which works well and fits the general aesthetic / workflow of `sbctl`